### ISSUE TITLE
com.jsoniter.spi.JsonException: no constructor for: interface

### ISSUE THREAD

**Issue by pasniak** [json-iterator/java#175]
Hi guys, 

I am trying serialize/deserialize* my application's State class which contains MutableList<IThing> with
SomeClass extending IThing
SomeOtherClass extending IThing
...etc...

So far deserialization gives me 
```com.jsoniter.spi.JsonException: no constructor for: interface```

Can you please advise how to approach that and point to some testcase processing a list of object inheriting from an interface?

*)
val serializedState = JsonStream.serialize(state)
...
 JsonIterator.deserialize(stateString, mymodel.State::class.java)

I do not have any jsoniter customizations (attributes) so far.

**Comment by taowen** [comment]
you need at least some field in the json to represent which concrete class to use. then we need to customize the jsoniter, such that the field will be serialized out, and used in deserialization.  there is no standard way.

**Comment by pasniak** [comment]
Thanks Taowen - I dropped inheritance and redid my class to a some sort of variant i.e. it contains all possible types as nullable members. The demo below (Kotlin) works fine, everything get reconstituted nicely. 

But in my Android app I observe a weird bug where the first and second object (Wpt) on the deserialized list have **identical** Location (those of the first Wpt, lat=40.77725) even if the debugger (and log) shows the JSON (from which they were created) has **2 different** locations (JSON below)!  

Is there some optimization kicking in which the JsonIterator.deserialize would create Location only once?  When I break in the Wpt constructor (in Android case), I see 2 constructor calls (from createNewObject); one with default values Location(0,0) and one with Location(40.77725,-73.872611).  In short, EWR location gets created but it has LGA locations' location!

```
class Apt
class Nav

class Wpt (
    val loc: Location = DefaultLocation,
    var apt: Apt? = null,
    var nav: Nav? = null
)

class Rte (var r: MutableList<Wpt> = mutableListOf())

class TestJsonSerialization {
    @Test
    fun testSerialize() {
        val r = Rte(mutableListOf(
            Wpt(loc = Location(-1.0,-1.0), apt = Apt()),
            Wpt(loc = Location(-2.0,-2.0), nav = Nav()))
        )
        val json = JsonStream.serialize(r)
        val deserialized = JsonIterator.deserialize(json, Rte::class.java )
```

```
{"isGpsAlive":true,"currentRoute":{"name":"current","waypoints":[{"isPassed":false,"apt":{},"fix":null,"id":"LGA","location":{"lat":40.77725,"lon":-73.872611},"nav":null,"user":null},{"isPassed":false,"apt":{},"fix":null,"id":"EWR","location":{"lat":40.69248,"lon":-74.168687},"nav":null,"user":null}]},"directionOfTravel":null,"lastLocation":null}
```
(in Android I have a bit more fields)

**Comment by taowen** [comment]
https://github.com/json-iterator/java/blob/master/src/main/java/com/jsoniter/spi/Extension.java you can use your own decoder to share instance.

check https://github.com/json-iterator/java/tree/master/src/main/java/com/jsoniter/extra
